### PR TITLE
Reductions of Transpose, Adjoint, PermutedDimsArray

### LINKED ISSUE #33029: Reductions of Transpose, Adjoint, PermutedDimsArray
Taking the `sum` of a transposed matrix is much slower than summing the parent. Should there perhaps be methods which unwrap this?
```julia
rr = rand(1000,1000);
@btime sum($rr, dims=1) # 284.119 μs (1 allocation: 7.94 KiB)

tt = transpose(rr);
@btime sum($tt, dims=2)' # 1.504 ms (7 allocations: 8.13 KiB)
```
Seems similar on Julia 1.0 to 1.3, and also for `adjoint` & `PermutedDimsArray`.

### PR DESCRIPTION
**Pull request by mcabbott**
Closes #33029.

~~It also adds some methods to `∘` such that `adjoint ∘ identity ∘ adjoint === identity`, since it will tend to produce such combinations, and something else may wish to dispatch on them. But it would work without this, I can separate it if that's preferable. (Perhaps we should also complete the group with `conj ∘ transpose = adjoint` etc.)~~ Now removed, in favour of an explicit `_sandwich` function. 

The reduction functions which this overloads for `PermutedDimsArray` were defined after its file is loaded. I moved this file later, but could equally well move these methods to another file.

### CODE REVIEW

File: stdlib/LinearAlgebra/src/adjtrans.jl [hunk 1]
```diff
@@ -285,6 +293,19 @@ Broadcast.broadcast_preserving_zero_d(f, avs::Union{Number,AdjointAbsVec}...) =
...
+
+Base._mapreduce_dim(f, op, init::Base._InitialValue, A::Transpose, dims::Colon) =
+    transpose(Base._mapreduce_dim(transpose ∘ f ∘ transpose, op, init, parent(A), dims))
+Base._mapreduce_dim(f, op, init::Base._InitialValue, A::Adjoint, dims::Colon) =
+    adjoint(Base._mapreduce_dim(adjoint ∘ f ∘ adjoint, op, init, parent(A), dims))
+
+Base.mapreducedim!(f, op, B::AbstractArray, A::TransposeAbsMat) =
+    transpose(Base.mapreducedim!(transpose ∘ f ∘ transpose, op, transpose(B), parent(A)))
+Base.mapreducedim!(f, op, B::AbstractArray, A::AdjointAbsMat) =
+    adjoint(Base.mapreducedim!(adjoint ∘ f ∘ adjoint, op, adjoint(B), parent(A)))
```
Comment (mcabbott) [inline stdlib/LinearAlgebra/src/adjtrans.jl:306]: Thinking some more, it's possible this needs something like `(x,y) -> op(x', y')'` for the reduction operator.

**Comment by mcabbott** [comment]
Bump. Perhaps paging @tkf, since you were involved in #33029

File: base/permuteddimsarray.jl [hunk 2]
```diff
@@ -253,6 +253,16 @@ end
     P
 end
 
+function Base._mapreduce_dim(f, op, init::Base._InitialValue, A::PermutedDimsArray, dims::Colon)
+    Base._mapreduce_dim(f, op, init, parent(A), dims)
+end
+
+function Base.mapreducedim!(f, op, B::AbstractArray{T,N}, A::PermutedDimsArray{T,N,P,Q}) where {T,N,P,Q}
+    C = PermutedDimsArray{T,N,Q,P,typeof(B)}(B)
```
Comment (vtjnash) [inline base/permuteddimsarray.jl:261]:
```suggestion
    C = PermutedDimsArray{T,N,P,Q,typeof(B)}(B)
```

Comment (vtjnash) [inline base/permuteddimsarray.jl:261]: EDIT: I think I understand why it is legal to swap these cases. It seems intuitive anyways, even I can't prove it. Can you change `P` and `Q` to `perm` and `iperm`, as I see them named elsewhere to make this clearer here?
```suggestion
function Base.mapreducedim!(f, op, B::AbstractArray{T,N}, A::PermutedDimsArray{T,N,perm,iperm}) where {T,N,perm,iperm}
    C = PermutedDimsArray{T,N,iperm,perm,typeof(B)}(B) # make the inverse permutation for the output
```

**Review by vtjnash — APPROVED** [review]

File: base/permuteddimsarray.jl (hunk 2 above)
Comment (mcabbott) [reply base/permuteddimsarray.jl:261]: The inverse here is essential, here's a check: 
```julia
X = rand(10,20,300);
A = permutedims(X, (2,3,1)); summary(A)
P = PermutedDimsArray(X, (2,3,1)); summary(P)

sum(P, dims=2) ≈ sum(A, dims=2)

@btime sum(P, dims=2); # 56.125 μs
@btime sum(A, dims=3); # 15.292 μs

function Base.mapreducedim!(f, op, B::AbstractArray{T,N}, A::PermutedDimsArray{T,N,P,Q}) where {T,N,P,Q}
    C = PermutedDimsArray{T,N,Q,P,typeof(B)}(B)
    Base.mapreducedim!(f, op, C, parent(A))
    B
end

sum(P, dims=2) ≈ sum(A, dims=2)
@btime sum(P, dims=2); # 24.666 μs

function Base.mapreducedim!(f, op, B::AbstractArray{T,N}, A::PermutedDimsArray{T,N,P,Q}) where {T,N,P,Q}
    C = PermutedDimsArray{T,N,P,Q,typeof(B)}(B)  # with suggestion
    Base.mapreducedim!(f, op, C, parent(A))
    B
end

sum(P, dims=2) ≈ sum(A, dims=2) # DimensionMismatch
```

Comment (mcabbott) [reply base/permuteddimsarray.jl:261]: Sure. I thought I was following Base, but in fact it seems I was following my package... will change.

**Comment by mcabbott** [comment]
This doesn't seem speed up `maximum`:
```julia
julia> X = randn(5000, 200);

julia> @btime maximum($X);
  636.708 μs (0 allocations: 0 bytes)

julia> @btime maximum(transpose($X));
  4.643 ms (0 allocations: 0 bytes)

julia> @btime maximum($X; dims=1);
  744.458 μs (4 allocations: 1.88 KiB)

julia> @btime maximum(transpose($X); dims=2);
  4.644 ms (14 allocations: 2.06 KiB)
```
although `sum` works as expected:
```
julia> @btime sum($X);
  200.250 μs (0 allocations: 0 bytes)

julia> @btime sum(transpose($X));
  200.250 μs (0 allocations: 0 bytes)
```

**Pull request merged**